File: linked-lists/6.c
/**
 * 6.c
 * Linked Lists
 * https://github.com/ashenm/dsa
 *
 * Implements Stack-like operations
 *
 * Ashen Gunaratne
 * mail@ashenm.ml
 *
 */

#include <stdio.h>
#include <stdlib.h>

typedef struct node
{
    int value;
    struct node *next;
} node;

// conjoins a new node
node* push (node **head, int value);

// removes the last insert
node* pop (node **head);

// computes the node count
int range (node *head);

int main (void)
{

    node *head = NULL;

    push(&head, 25);
    push(&head, 50);

    printf("%i\n", range(head));

    free(pop(&head));

    node *elect = head;

    while (elect != NULL)
    {
        printf("%i\n", elect -> value);
        elect = elect -> next;
    }

    while (head != NULL)
    {
        elect = head;
        head = head -> next;
        free(elect);
    }

}

node* push (node **head, int value)
{

    node *elect = *head;
    node *inset = malloc(sizeof(node));

    if (inset == NULL)
    {
        return NULL;
    }

    inset -> value = value;
    inset -> next = elect;

    return *head = inset;

}

node* pop (node **head)
{

    node *elect = *head;

    if (elect == NULL)
    {
        return NULL;
    }

    *head = elect -> next;

    return elect;

}

int range (node *head)
{

    int n = 0;

    while (head != NULL)
    {
        head = head -> next;
        n += 1;
    }

    return n;

}

// vim: set expandtab shiftwidth=4:

File: linked-lists/6.js
/**
 * 6.js
 * Linked Lists
 * https://github.com/ashenm/dsa
 *
 * Implements Stack-like operations
 *
 * Ashen Gunaratne
 * mail@ashenm.ml
 *
 */

'use strict';

class Node {

  constructor (value, next=null) {
    this.value = value;
    this.next = next;
  }

}

class LinkedList {

  constructor (head=null) {
    this.head = head;
  }

  // conjoins a new node
  push (value) {
    return this.head = new Node(value, this.head), this.head;
  }

  // removes the last insert
  pop () {

    const elect = this.head;

    if (elect) {
      this.head = elect.next;
    }

    return elect;

  }

  // computes the node count
  range () {

    let n = 0;
    let elect = this.head;

    while (elect) {
      elect = elect.next;
      n += 1;
    }

    return n;

  }

}

const main = function () {

  const list = new LinkedList();

  list.push(25);
  list.push(50);

  console.log(list.range());

  list.pop();

  for (let elect = list.head; elect; elect = elect.next) {
    console.log(elect.value);
  }

};

if (require.main === module) {
  main();
}

// vim: set expandtab shiftwidth=2:
